rnwinrt/rnwinrt/react/strings/base.cpp

```diff
@@ -205,22 +205,29 @@ jsi::Value static_enum_data::create(jsi::Runtime& runtime) const
     return jsi::Value(runtime, jsi::Object::createFromHostObject(runtime, std::make_shared<projected_enum>(this)));
 }
 
-jsi::Value static_enum_data::get_value(std::string_view valueName) const
+jsi::Value static_enum_data::get_value(jsi::Runtime& runtime, std::string_view valueName) const
 {
     // TODO: It would also be rather simple to ensure that the array is sorted and do a binary search, however the
     // number of enum elements is typically pretty small, so that may actually be harmful
-    auto itr = std::find_if(values.begin(), values.end(), [&](auto& mapping) { return mapping.name == valueName; });
+    auto itr = std::find_if(values.begin(), values.end(),
+        [&](auto& mapping) { return mapping.name == valueName || std::to_string((long)mapping.value) == valueName; });
```

Comment:
I'm not too keen on doing a `std::to_string` on every iteration of the loop. A couple alternate suggestions:

1. Modify the code generation so that `mapping` contains the name, value, and a string representation of the value
2. Modify this code to do a string search only if `find_if` returns the end iterator

Reply:
Perhaps more broadly, is there a particular reason why enum projections (in their current implementation) seem to disregard the original underlying type from the WinMD?

We could get rid of any sort of assumptions if we instead were to implement enum projections in a way that preserved the original underlying type ... something like:

```cpp
template<typename UnderlyingType>
struct static_enum_data final : static_projection_data
{
    struct value_mapping
    {
        std::string_view name;
        UnderlyingType value;
    }

    // Implementations for this struct's methods would need to move from the cpp into here
    ...
}
```

Where the [codegen on enums ](https://github.com/microsoft/react-native-winrt/blob/9b8b4db28e4dcbd72c405e2e1603116ec6c20a98/rnwinrt/rnwinrt/react/ReactFileGenerator.cpp#L400)could do something like:

```cpp
static void write_rnwinrt_enum_projection_data(rnwinrt::writer& writer, const enum_projection_data& enumData)
{
    ...
    const auto underlyingType = enumData.type_def.get_enum_definition().m_underlying_type
    for (auto&& field : enumData.type_def.FieldList())
    {
        if (!field.Constant()) // Enums have a 'value__' data member we need to ignore
            continue;

        writer.write_fmt(R"^-^(
        { "%"sv, static_cast<%>(winrt::%::%) },)^-^",
            rnwinrt::camel_case{ field.Name() },
            [&](rnwinrt::writer& w) { rnwinrt::write_cppwinrt_type(w, underlyingType); },
            rnwinrt::cpp_typename{ enumData.type_def },
            field.Name());
    }
    ...
}
```

Then searching can happen in the way you suggested, with greater guarantees for the reverse-mapping part by using the underlying type knowledge:

```cpp
jsi::Value get_value(std::string_view valueName)
{
    auto itr = std::find_if(values.begin(), values.end(), [&](auto& mapping) { return mapping.name == valueName; });
    if (itr == values.end())
    {
        if (std::is_same_v<UnderlyingType, int32_t>)
        {
            // Parse valueName into an int32_t, then search by value
        }
        ...
    }   
}
```

Reply:
> is there a particular reason why enum projections (in their current implementation) seem to disregard the original underlying type from the WinMD

Because ultimately these values need to be reflected back into JavaScript as a `number` (i.e. an IEEE 64-bit floating point). There's limited value to retaining knowledge of the underlying type, and my initial suggestion was misguided to suggest it. Given that enum values are guaranteed to be 32-bits, and given that all 32-bit integers - signed or unsigned - can be represented by a `double`, we don't lose anything by just using it here. Given that, I think this should look something like the following (example; I didn't check to verify that it compiles):

```c++
jsi::Value static_enum_data::get_value(std::string_view valueName) const
{
    // TODO: It would also be rather simple to ensure that the array is sorted and do a binary search, however the
    // number of enum elements is typically pretty small, so that may actually be harmful
    auto itr = std::find_if(values.begin(), values.end(), [&](auto& mapping) { return mapping.name == valueName; });
    if (itr != values.end())
    {
        return jsi::Value(itr->value);
    }

    double valueNumber;
    auto [ptr, ec] = std::from_chars(valueName.c_str(), valueName.c_str() + valueName.size(), valueNumber);
    if ((ec == std::errc{}) && (ptr == (valueName.c_str() + valueName.size()))
    {
        itr = std::find_if(values.begin(), values.end(), [&](auto& mapping) { return mapping.value == valueNumber; });
        if (itr != values.end())
        {
            return jsi::Value(itr->value);
        }
    }

    return jsi::Value::undefined();
}
```

Reply:
> 2. Modify this code to do a string search only if `find_if` returns the end iterator. If we go this route, we should actually do the opposite: attempt to convert `valueName` to `double`. If that succeeds, we do the search

I've made the first part of this change. I did not make the second part given that JavaScript does not do that kind of matching normally.